src/Compilers/CSharp/Portable/Symbols/Source/SourceComplexParameterSymbol.cs

```diff
@@ -194,35 +193,48 @@ internal bool HasEnumeratorCancellationAttribute
+                if (state.NotePartComplete(CompletionPart.StartDefaultSyntaxValue))
                 {
                     var diagnostics = DiagnosticBag.GetInstance();
-
-                    if (Interlocked.CompareExchange(
+                    var previousValue = Interlocked.CompareExchange(
                         ref _lazyDefaultSyntaxValue,
                         MakeDefaultExpression(diagnostics, out var binder, out var parameterEqualsValue),
-                        ConstantValue.Unset) == ConstantValue.Unset)
-                    {
-                        if (binder is not null && parameterEqualsValue is not null && !_lazyDefaultSyntaxValue.IsBad)
-                        {
-                            NullableWalker.AnalyzeIfNeeded(binder, parameterEqualsValue, diagnostics);
-                        }
-                        NullableAnalyzeParameterDefaultValueFromAttributes(diagnostics);
+                        ConstantValue.Unset);
+                    Debug.Assert(previousValue == ConstantValue.Unset);
+
+                    var completedOnThisThread = state.NotePartComplete(CompletionPart.EndDefaultSyntaxValue);
+                    Debug.Assert(completedOnThisThread);
 
-                        // This is a race condition where the thread that loses
-                        // the compare exchange tries to access the diagnostics
-                        // before the thread which won the race finishes adding
-                        // them. https://github.com/dotnet/roslyn/issues/17243
-                        AddDeclarationDiagnostics(diagnostics);
+                    if (binder is not null && parameterEqualsValue is not null && !_lazyDefaultSyntaxValue.IsBad)
+                    {
+                        NullableWalker.AnalyzeIfNeeded(binder, parameterEqualsValue, diagnostics);
+                        VerifyParamDefaultValueMatchesAttributeIfAny(_lazyDefaultSyntaxValue, parameterEqualsValue.Value.Syntax, diagnostics);
                     }
 
+                    AddDeclarationDiagnostics(diagnostics);
                     diagnostics.Free();
+
+                    completedOnThisThread = state.NotePartComplete(CompletionPart.EndDefaultSyntaxValueDiagnostics);
+                    Debug.Assert(completedOnThisThread);
                 }
 
+#if DEBUG
+                // If we accidentally access DefaultSyntaxValue in a reentrant manner, and we use a cancellation token
+                // that never expires, then the failure mode will be to spin wait forever.
+                // For debug purposes let's instead use a token which expires after a modest amount of time
+                // to wait for the default syntax value to be available.
+                var cts = new CancellationTokenSource(TimeSpan.FromSeconds(5));
+                var token = cts.Token;
+#else
+                var token = CancellationToken.None;
+#endif
+                state.SpinWaitComplete(CompletionPart.EndDefaultSyntaxValue, token);
```

Comment:
Feel like `ForceComplete` should change to assert that this part has completed. That is the pattern we usually have elsewhere.

Reply:
I didn't find examples of asserting that parts are completed instead of spin-waiting for them to complete when looking through the overrides of [Symbol.ForceComplete](https://github.com/dotnet/roslyn/blob/42629354bc746019e212df6a0a991aeb75d12282/src/Compilers/CSharp/Portable/Symbols/Symbol.cs#L49). Could you provide an example of what you had in mind?